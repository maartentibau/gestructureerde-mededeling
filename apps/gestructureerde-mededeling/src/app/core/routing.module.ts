import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { CreateComponent } from '../components/create/create.component';
import { GenerateComponent } from '../components/generate/generate.component';
import { ValidateComponent } from '../components/validate/validate.component';

const routes: Routes = [
  { path: 'create', component: CreateComponent },
  { path: 'generate', component: GenerateComponent },
  { path: 'validate', component: ValidateComponent },
  { path: '', redirectTo: '/generate', pathMatch: 'full' },
  { path: '**', component: GenerateComponent },
];

@NgModule({
  imports: [RouterModule.forRoot(routes, { relativeLinkResolution: 'legacy' })],
  exports: [RouterModule],
})
export class RoutingModule {}
