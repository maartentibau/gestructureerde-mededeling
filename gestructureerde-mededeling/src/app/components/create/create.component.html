<ogm-number [ogm]="ogm$ | async"></ogm-number>

<ogm-input
  (ogmInputChange)="ogmInputChangeHandler($event)"
  placeholderMessage="vul een getal in van maximaal 10 cijfers"
></ogm-input>

<ogm-controls
  *ngIf="ogm"
  [ogm]="ogm"
  [copyNumber]="true"
  [copyOgm]="true"
  (copyNumberClick)="copyNumberClickHandler($event)"
  (copyOgmClick)="copyOgmClickHandler($event)"
></ogm-controls>
