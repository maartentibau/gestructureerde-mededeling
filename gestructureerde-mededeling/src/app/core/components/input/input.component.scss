@import '../../../../assets/styles/screen-media-queries';

.ogm-form {
  text-align: center;
  padding: 0 10%;
  display: flex;
  margin: 1rem 0;
}

input {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  flex: 1 1 auto;
  text-align: center;
  display: block;
  padding: 8px 24px;
  border-radius: 4px;
  border: 1px solid #e9e9e9;
  font-size: 250%;
  box-shadow: 0px 2px 1px -1px rgba(0, 0, 0, 0.2), 0px 1px 1px 0px rgba(0, 0, 0, 0.14),
    0px 1px 3px 0px rgba(0, 0, 0, 0.12);
  outline: none;
  color: #103183;

  &::placeholder {
    color: #e1e1e1;
  }

  @media #{$mat-lt-lg} {
    font-size: 180%;
  }

  @media #{$mat-lt-md} {
    font-size: 120%;
  }

  @media #{$mat-lt-sm} {
    font-size: 100%;
  }

  @media #{$mat-lt-xs} {
    font-size: 90%;
  }
}
