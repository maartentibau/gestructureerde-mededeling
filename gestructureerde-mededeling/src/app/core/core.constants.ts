export const DEFAULT_TITLE = `Gestructureerde Mededeling - OGM`;
