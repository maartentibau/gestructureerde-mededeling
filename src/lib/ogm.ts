export function ogm(): string {
  return 'ogm';
}
