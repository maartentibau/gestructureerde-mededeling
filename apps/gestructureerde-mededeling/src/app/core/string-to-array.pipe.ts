import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
  name: 'stringToArray',
})
export class StringToArrayPipe implements PipeTransform {
  transform(value: string | null | undefined): string[] {
    if ((value && typeof value === 'string') || typeof value === 'number') {
      return Array.from(String(value));
    }

    return [];
  }
}
