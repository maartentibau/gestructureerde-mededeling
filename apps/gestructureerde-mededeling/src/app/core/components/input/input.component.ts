import { Component, EventEmitter, Input, OnDestroy, OnInit, Output } from '@angular/core';
import { FormControl } from '@angular/forms';
import { Observable, Subject } from 'rxjs';
import { filter, map, takeUntil, tap } from 'rxjs/operators';
import { OgmService } from '../../services/ogm.service';

export interface OgmInputChange {
  ogm: string;
  isValid: boolean;
}

@Component({
  selector: 'ogm-input',
  templateUrl: './input.component.html',
  styleUrls: ['./input.component.scss'],
})
export class InputComponent implements OnInit, OnDestroy {
  @Input() validate: boolean;
  @Input() placeholderMessage: string;

  @Output() ogmInputChange: EventEmitter<OgmInputChange> = new EventEmitter<undefined>();

  readonly ogmInput: FormControl;
  readonly destroy$: Subject<void>;

  ogm$: Observable<string>;

  constructor(private ogmService: OgmService) {
    this.ogmInput = new FormControl(null);
    this.destroy$ = new Subject<void>();
  }

  ngOnInit() {
    const ogmInitValue: string = this.ogmService.init();
    const hasSpaces: RegExp = new RegExp('[\\s]');

    const omgInputChanges$ = this.ogmInput.valueChanges.pipe(
      filter((value) => {
        if (isNaN(Number(value)) || hasSpaces.test(value)) {
          const cleanValue = this.ogmService.clean(value);
          this.ogmInput.setValue(cleanValue, { emitEvent: false });
          return false;
        }

        return true;
      })
    );

    if (this.validate) {
      this.ogm$ = omgInputChanges$.pipe(
        map((value) => (value ? this.ogmService.format(value.toString().padEnd(12, ' ')) : ogmInitValue))
      );
    } else {
      this.ogm$ = omgInputChanges$.pipe(map((value) => (value ? this.ogmService.generate(value, true) : ogmInitValue)));
    }

    this.ogm$
      .pipe(
        tap((ogm: string) => {
          let isValid;

          if (this.validate) {
            const ogmNumber = this.ogmService.clean(ogm);
            isValid = ogmNumber.length === 12 ? this.ogmService.validate(ogm) : undefined;
          }

          this.ogmInputChange.emit({ ogm, isValid });
        }),
        takeUntil(this.destroy$)
      )
      .subscribe();
  }

  ngOnDestroy() {
    this.destroy$.next();
  }
}
