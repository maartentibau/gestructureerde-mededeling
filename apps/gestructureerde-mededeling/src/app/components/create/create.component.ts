import { ChangeDetectionStrategy, Component } from '@angular/core';
import { Title } from '@angular/platform-browser';
import { BehaviorSubject } from 'rxjs';
import { DEFAULT_TITLE } from '../../core/core.constants';
import { OgmService } from '../../core/services/ogm.service';
import { OgmInputChange } from '../../core/components/input/input.component';
import { MatSnackBar } from '@angular/material/snack-bar';
import { OgmData } from '../../core/ogm.model';

@Component({
  selector: 'ogm-create',
  templateUrl: './create.component.html',
  styleUrls: ['./create.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CreateComponent {
  readonly ogm$: BehaviorSubject<string>;

  ogm: OgmData | undefined;

  constructor(private ogmService: OgmService, private snackBar: MatSnackBar, private title: Title) {
    this.title.setTitle(`${DEFAULT_TITLE} - Zelf een gestructureerde mededeling maken`);

    this.ogm$ = new BehaviorSubject<string>(this.ogmService.init());
  }

  ogmInputChangeHandler({ ogm }: OgmInputChange) {
    const cleanOgm = this.ogmService.clean(ogm);

    this.ogm = cleanOgm === '' ? undefined : { number: cleanOgm, numberFormat: ogm };
    this.ogm$?.next(ogm);
  }

  copyNumberClickHandler(message: string) {
    this.showSnackbarMessage(message);
  }

  copyOgmClickHandler(message: string) {
    this.showSnackbarMessage(message);
  }

  showSnackbarMessage(message?: string) {
    if (message) {
      this.snackBar.open(message, '', {
        panelClass: 'center-clipboard-text',
        verticalPosition: 'top',
        duration: 750,
      });
    }
  }
}
