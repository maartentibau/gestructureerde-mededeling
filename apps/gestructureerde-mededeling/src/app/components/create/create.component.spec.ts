import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { Title } from '@angular/platform-browser';
import { DEFAULT_TITLE } from '../../core/core.constants';

import { CreateComponent } from './create.component';
import { OgmService } from '../../core/services/ogm.service';
import { Component, Input } from '@angular/core';
import { OgmInputChange } from '../../core/components/input/input.component';
import { MatSnackBar } from '@angular/material/snack-bar';
import { MaterialModule } from '../../core/material.module';
import { OgmData } from '../../core/ogm.model';

@Component({
  selector: 'ogm-number',
  template: `
    <div>{{ ogm | json }}</div>
    <div>{{ isValid | json }}</div>
  `,
})
class MockNumberComponent {
  @Input() ogm: string;
  @Input() isValid: boolean;
}

@Component({
  selector: 'ogm-input',
  template: `
    <div>{{ validate | json }}</div>
    <div>{{ placeholderMessage | json }}</div>
  `,
})
class MockInputComponent {
  @Input() validate: boolean;
  @Input() placeholderMessage: string;
}

@Component({
  selector: 'ogm-controls',
  template: `
    <div>{{ refresh | json }}</div>
    <div>{{ copyNumber | json }}</div>
    <div>{{ copyOgm | json }}</div>
    <div>{{ ogm | json }}</div>
  `,
})
class MockControlsComponent {
  @Input() refresh: boolean;
  @Input() copyNumber: boolean;
  @Input() copyOgm: boolean;
  @Input() ogm: OgmData;
}

describe('CreateComponent', () => {
  let component: CreateComponent;
  let fixture: ComponentFixture<CreateComponent>;
  let snackBar: MatSnackBar;
  let ogmService: OgmService;
  let titleService: Title;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [MaterialModule],
      declarations: [CreateComponent, MockNumberComponent, MockInputComponent, MockControlsComponent],
      providers: [OgmService, MatSnackBar],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CreateComponent);
    component = fixture.componentInstance;

    snackBar = TestBed.inject(MatSnackBar);
    ogmService = TestBed.inject(OgmService);
    titleService = TestBed.inject(Title);
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('rendering component', () => {
    it('should match snapshot', () => {
      // act
      fixture.detectChanges();

      // check
      expect(fixture).toMatchSnapshot();
    });
  });

  describe('constructor', () => {
    it('set the correct HTML title', () => {
      // prepare
      const title: string = `${DEFAULT_TITLE} - Zelf een gestructureerde mededeling maken`;
      jest.spyOn(titleService, 'setTitle');

      // act
      TestBed.createComponent(CreateComponent);

      // check
      expect(titleService.setTitle).toHaveBeenCalledWith(title);
    });
  });

  describe('ogmInputChangeHandler', () => {
    it('should set ogm value to undefined and call next on ogm$ stream with init numberFormat', () => {
      // prepare
      const ogm: OgmData = undefined;
      jest.spyOn(component.ogm$, 'next');

      const ogmInputChange: OgmInputChange = { ogm: '+++   /    /     +++', isValid: undefined };

      // act
      component.ogmInputChangeHandler(ogmInputChange);

      // check
      expect(component.ogm).toBeUndefined();
      expect(component.ogm$.next).toHaveBeenCalledWith(ogmInputChange.ogm);
    });

    it('should set ogm with the correct data and call next on ogm$ stream with the numberFormat', () => {
      // prepare
      const ogm: OgmData = { number: '120000000002', numberFormat: '+++120/0000/00002+++' };
      jest.spyOn(component.ogm$, 'next');

      const ogmInputChange: OgmInputChange = { ogm: ogm.numberFormat, isValid: true };

      // act
      component.ogmInputChangeHandler(ogmInputChange);

      // check
      expect(component.ogm).toStrictEqual(ogm);
      expect(component.ogm$.next).toHaveBeenCalledWith(ogmInputChange.ogm);
    });
  });

  describe('Copy number or OGM', () => {
    beforeEach(() => {
      jest.spyOn(component, 'showSnackbarMessage');
    });

    describe('copyNumberClickHandler', () => {
      it('should the correct copy message when called', () => {
        // prepare
        const message: string = 'Copy number!';

        // act
        component.copyNumberClickHandler(message);

        // check
        expect(component.showSnackbarMessage).toHaveBeenCalledWith(message);
      });
    });

    describe('copyOgmClickHandler', () => {
      it('should the correct copy message when called', () => {
        // prepare
        const message: string = 'Copy OGM!';

        // act
        component.copyOgmClickHandler(message);

        // check
        expect(component.showSnackbarMessage).toHaveBeenCalledWith(message);
      });
    });
  });
});
