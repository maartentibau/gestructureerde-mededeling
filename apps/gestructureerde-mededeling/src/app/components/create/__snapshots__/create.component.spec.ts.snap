// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateComponent rendering component should match snapshot 1`] = `
<ogm-create
  __ngContext__={[Function LRootView]}
  ogm$={[Function BehaviorSubject]}
  ogmService={[Function OgmService]}
  snackBar={[Function MatSnackBar]}
  title={[Function Title]}
>
  <ogm-number>
    <div>
      "+++   /    /     +++"
    </div>
    <div>
      
    </div>
  </ogm-number><ogm-input
    placeholdermessage="vul een getal in van maximaal 10 cijfers"
  >
    <div>
      
    </div>
    <div>
      "vul een getal in van maximaal 10 cijfers"
    </div>
  </ogm-input>
</ogm-create>
`;
